### PR TITLE
Add redfish BMC support in baremetal operator

### PR DESCRIPTION
**Pull request by maelk**
accepts bmc address like
`redfish+https://my.host:8080/redfish/v1/Systems/MySystem`

### CODE REVIEW

File: pkg/bmc/redfish.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+func (a *redfishAccessDetails) DriverInfo(bmcCreds Credentials) map[string]interface{} {
+	redfishAddress := []string{}
+	schemes := strings.Split(a.bmcType, "+")
+	if len(schemes) > 1 {
+		redfishAddress = append(redfishAddress, schemes[1])
+	} else {
+		redfishAddress = append(redfishAddress, redfishDefaultScheme)
+	}
+	redfishAddress = append(redfishAddress, "://")
+	redfishAddress = append(redfishAddress, a.hostname)
```
Comment (dtantsur) [inline pkg/bmc/redfish.go]: IIRC (@etingof could confirm) the `/redfish/v1` bit has to be present in both `redfishAddress` and `redfish_system_id`.

File: pkg/bmc/redfish.go [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+// NeedsMAC returns true when the host is going to need a separate
+// port created rather than having it discovered.
+func (a *redfishAccessDetails) NeedsMAC() bool {
+	// libvirt-based hosts used for dev and testing require a MAC
+	// address, specified as part of the host, but we don't want the
+	// provisioner to have to know the rules about which drivers
+	// require what so we hide that detail inside this class and just
+	// let the provisioner know that "some" drivers require a MAC and
+	// it should ask.
+	return false
```
Comment (dtantsur) [inline pkg/bmc/redfish.go]: Must be `true`, inspection lookup won't work otherwise.

File: pkg/bmc/redfish.go (hunk 1 above)
Comment (maelk) [reply pkg/bmc/redfish.go]: We used this as a base for implementation: https://docs.openstack.org/ironic/stein/admin/drivers/redfish.html There the redfish/v1 does not seem to be needed for redfish_address.

File: pkg/bmc/redfish.go (hunk 2 above)
Comment (maelk) [reply pkg/bmc/redfish.go]: Could you please explain this ? From the ipmi.go, it returns `return a.bmcType == "libvirt"` meaning that when we use "ipmi://xxx" addresses such as dev env, it returns false. And what about libvirt/ non-libvirt based hosts ?

Comment (dhellmann) [reply pkg/bmc/redfish.go]: I think one of the goals is to replace the current IPMI-based dev environment with one that uses Redfish. That would mean using the Redfish driver when the type is "libvirt". And that would mean this function should have logic similar to the current IPMI implementation, at least.

Unless @dtantsur's comment means that all Redfish servers always need a MAC pre-defined in order to have their boot process managed?

Comment (dtantsur) [reply pkg/bmc/redfish.go]: To be clear, PXE booting requires MAC address to be present. We can use introspection to discover NICs (and thus MAC addresses), however, it has a chicked-and-egg problems: without both MAC addresses and IPMI address in our records, how do we know which introspection data corresponds to which node?

TL;DR Redfish requires a MAC address with and without inspection for the time being.

File: pkg/bmc/redfish.go (hunk 1 above)
Comment (dtantsur) [reply pkg/bmc/redfish.go]: You're probably right, sorry for confusion.

File: pkg/bmc/redfish.go (hunk 2 above)
Comment (dhellmann) [reply pkg/bmc/redfish.go]: @dtantsur why is that different for redfish from IPMI (which only requires a MAC for VMs)?

Comment (maelk) [reply pkg/bmc/redfish.go]: Aren't we using the IPMI driver for dev env (and not Libvirt) : https://github.com/metal3-io/metal3-dev-env/blob/master/vm-setup/roles/libvirt/templates/ironic_nodes.json.j2#L15 And in that case the needsMAC returns false. So how does the dev env works if the MAC should be required ?

Comment (maelk) [reply pkg/bmc/redfish.go]: Aren't we using the IPMI driver for dev env (and not Libvirt) : https://github.com/metal3-io/metal3-dev-env/blob/master/vm-setup/roles/libvirt/templates/ironic_nodes.json.j2#L15 then used here : https://github.com/metal3-io/metal3-dev-env/blob/master/lib/common.sh#L133 . And in that case the needsMAC returns false. So how does the dev env works if the MAC should be required ?

Comment (maelk) [reply pkg/bmc/redfish.go]: and what if we switched from PXE to redfish-virtual-media boot interface ?

Comment (dhellmann) [reply pkg/bmc/redfish.go]: The scripts for building the dev environment populate the MAC address, so even if the driver isn't requiring it, it's there.

Comment (dtantsur) [reply pkg/bmc/redfish.go]: > why is that different for redfish from IPMI (which only requires a MAC for VMs)?

Because on real hardware we can detect IPMI address from within the machine and use it for matching data to nodes. It's much more complex for Redfish.

> what if we switched from PXE to redfish-virtual-media boot interface ?

Might work, needs testing.

Comment (dtantsur) [reply pkg/bmc/redfish.go]: > What additional work would be needed in this repo to make it work with ironic configured that way?

To clarify on my comment: inspection is independent of the boot method in ironic, it always uses PXE and requires a port or an IPMI address for matching. Virtual media only helps if we don't use inspection.

Comment (maelk) [reply pkg/bmc/redfish.go]: ok thank you @dtantsur and @dhellmann . This is now fixed.

**Comment by maelk** [comment]
@dtantsur @dhellmann Any further comments ?

File: pkg/bmc/redfish.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+func newRedfishAccessDetails(bmcType, portNum, hostname, path string) (AccessDetails, error) {
+	// If the hostname is an ipv6 address, it needs to be in square brackets
+	// as we are forming a URL out of it.
+	addresses, err := net.LookupIP(hostname)
+	if err == nil {
+		for _, address := range addresses {
+			if address.String() == hostname && address.To4() == nil {
+				hostname = strings.Join([]string{"[",hostname,"]"}, "")
```
Comment (dhellmann) [inline pkg/bmc/redfish.go]: What happens here if a host has multiple IPv6 addresses?

Comment (dtantsur) [reply pkg/bmc/redfish.go]: Why do we even resolve the address?

Comment (maelk) [reply pkg/bmc/redfish.go]: Resolving the hostname is to find out if the domain name given is an ipv6 (that will need brackets around) or not. If we are given a domain name and the host has multiple addresses, then none will match the domain name. If we are given an IPv6, the resolved IP will match the given IP and the brackets will be added.

Comment (dhellmann) [reply pkg/bmc/redfish.go]: OK, the purpose of that loop was not obvious to me at all, even with the comment.

I think if we take the other PR that changes the access API to use a parsed URL, we would not need to perform this check here, right?

File: pkg/bmc/redfish.go [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+func init() {
+	registerFactory("redfish", newRedfishAccessDetails)
+	registerFactory("redfish+http", newRedfishAccessDetails)
+	registerFactory("redfish+https", newRedfishAccessDetails)
+}
+
+func newRedfishAccessDetails(bmcType, portNum, hostname, path string) (AccessDetails, error) {
+	// If the hostname is an ipv6 address, it needs to be in square brackets
+	// as we are forming a URL out of it.
+	addresses, err := net.LookupIP(hostname)
```
Comment (zaneb) [inline pkg/bmc/redfish.go]: It seems like you should be using `net.ParseIP()`, not `net.LookupIP()` here.

File: pkg/bmc/redfish.go (hunk 3 above)
Comment (zaneb) [reply pkg/bmc/redfish.go]: Correct, with that other PR you could just use `.Host` instead of `.Hostname()` and you get the unmodified host+port.
This seems like a good reason to accept that PR.

Comment (maelk) [reply pkg/bmc/redfish.go]: Ok, so we'll first focus on URL PR and then modify this one

**Comment by maelk** [comment]
now built on the URL PR, we got rid of the lookup part.

**Comment by maelk** [comment]
running the integration CI loop would not test anything from this change as metal3-dev-env uses IPMI. A PR to use Redfish for one vm is under work

**Comment by dhellmann** [comment]
> running the integration CI loop would not test anything from this change as metal3-dev-env uses IPMI. A PR to use Redfish for one vm is under work

If it's easier to just switch them all then I think a patch that added a variable to control that would be very welcome.

**Comment by maelk** [comment]
Any further comments ? (@zaneb ? )
